docker_container: ambiguous parameter "stop_timeout"

stromvirvel opened the pull request:
##### SUMMARY
Fixes #43814
Honour stop_timeout when creating docker containers and updates documentation accordingly.

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
docker_container

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.4
  config file = /Users/remo.wenger/.ansible.cfg
  configured module search path = [u'/Users/remo.wenger/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/Cellar/ansible/2.5.4/libexec/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 13:05:56) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

stromvirvel committed:
docker_container: Honour stop_timeout when creating docker containers (#43814)

kassiansun commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -363,7 +363,10 @@
   stop_timeout:
     description:
-      - Number of seconds to wait for the container to stop before sending SIGKILL.
+      - 'When C(state) is I(present), set container\'s C(StopTimeout) configuration (in seconds). This only applies on container creation time.
```
I think we should not use list here, just a multiple-line documentation like `command` option.
And the cause of sanity check may be the "'", just omit it.

kassiansun replied:
So the original `Number of seconds to wait for the container to stop before sending SIGKILL.` can be preserved as an intro for later two scenarios.

kassiansun commented:
shipit

moepot commented:
@kassiansun The bot did not labelled as `shipit`. I just read the Ansibullbot workflow and I guess I should have said `ready_for_review` before your `shipit` ... ?

kassiansun commented:
I think we need two shipit from maintainers to label it. But actually there's only one maintainer here, haha.

kassiansun commented:
@Akasurde The `docker_*` modules are now community-maintained, but there's not enough maintainers here.
I think @felixfontein is a good candidate, he has contributed quite some patches and reviews on `docker_*`. Would you mind being a maintainer for `docker_*` modules, @felixfontein ?

felixfontein commented:
Getting only one maintainer shipit for `docker_*` modules seems to be quite common nowadays. Most of my docker_container PRs are still in that state...
@kassiansun I wouldn't mind helping out, at least for `docker_container` (I don't really know most of the other modules, and never used them except `docker_image` once for pulling images, so I don't think I should be a maintainer for them).
@pilou- might be a good candidate as well, he did some work on the `docker_` modules as well (like fixing the `docker_container` init bug, and [working on tests](https://github.com/ansible/ansible/pull/42306)).

felixfontein commented:
@kassiansun If you want to add me as a maintainer, you can create a PR for [.github/BOTMETA.yml](https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml#L225).

Also, I understood the ansibot workflow that instead of two maintainers, one maintainer `shipit` and one other `shipit` should be enough. But apparently that doesn't work, see #16748...

kassiansun commented:
@felixfontein Per [ansibullbot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md#when-will-your-pull-request-be-merged), we need three if reviewer is not maintainer:

> at least three people which aren't maintainer nor core team member have approved the pull request using the shipit

felixfontein commented:
@kassiansun Ooops, I think I missed the "when both conditions are met"... So two more non-maintainer `shipits` would be needed.

felixfontein commented on lib/ansible/modules/cloud/docker/docker_container.py:
```diff
@@ -843,6 +847,7 @@ def create_parameters(self):
             mac_address='mac_address',
             labels='labels',
             stop_signal='stop_signal',
+            stop_timeout='stop_timeout',
```
Support for `stop_timeout` has only been added in `docker` 2.1.0 (see https://docker-py.readthedocs.io/en/stable/change-log.html). Since we support older versions, you need to add some more code which handles that:
  1) make sure this is only inserted into `create_params` when the `docker` client library version is high enough;
  2) raise an error if this parameter has a non-trivial value and version is too low.
Check the code of `AnsibleDockerClientContainer` (below in this file) for examples on how this can be done.

felixfontein commented:
@moepot Can you rebase this against current `devel`? Once that's done, we can also help you with the missing parts (by creating PRs against your branch). (Also, the current `devel` version now has tests!)

moepot commented:
@felixfontein rebase done. Sorry for my late reply. I'd be happy if someone could help me with the missing parts. 😊

felixfontein committed:
Adjusting description to what actually happens.

See docker-py changelog for 2.7.0: 'APIClient.stop will
no longer override the stop_timeout value present in the
container’s configuration.'

felixfontein committed:
Add a test whether stop_timeout can be configured for the container.

felixfontein committed:
Integrate with comparisons (by default, ignore stop_timeout value for restarts; will be configurable with PR ansible/ansible#44789).

felixfontein committed:
Added changelog.

felixfontein commented:
@moepot Now that #44789 has been merged, can you rebase another time? It should not make a difference, but I want to avoid misbehavior especially if moepot/ansible#1 is included in this PR.

moepot commented:
@felixfontein sure. I hope I made it right ...
Thanks for your precious help.

felixfontein commented:
@moepot I think so! Now let's see whether the integration tests still pass ;)

felixfontein commented:
BTW, I also tried to get docker-py to update `stop_timeout` on existing containers this morning, but wasn't able to get it done. Is that the reason why you wrote "only affects newly created containers"?

felixfontein commented:
The current CI failures are related to the changes in #44789 in combination to the changes here. I'll take a look at it later and create another PR :)

felixfontein commented:
With moepot/ansible#2 the integration tests should pass.

felixfontein committed:
Fix config change code and tests (#2)

* Improving wildcard test.
* Using correct config.

moepot commented:
@felixfontein yes exactly, it won't change the `stop_timeout` setting on existing containers. I'm fine with that as long as the docs declares it.

I've merged your PR. Do you need another rebase?

felixfontein reviewed (approved):
shipit

felixfontein commented:
@moepot Thanks! No, I think it's good now. The last rebase was needed because #44789 was merged today :)

@kassiansun Can you take a look at this again? I think it should be ready now.

kassiansun commented:
shipit

gundalow commented:
Merged into `devel` for release in Ansible 2.8

Thank you for the PR and reviewa

felixfontein commented:
@gundalow Thanks for merging this!
@moepot And thanks for staying with us and helping to get this done! Sorry that it took so long.
@kassiansun Thanks for the review!
